src/lib.rs

```diff
@@ -0,0 +1,350 @@
+extern crate core;
+
+pub mod commands;
+
+use crate::commands::CargoCommand;
+use cargo_metadata::{Message, MetadataCommand};
+use core::fmt;
+use rustc_version::Channel;
+use semver::Version;
+use serde::Deserialize;
+use std::io::{BufRead, BufReader};
+use std::path::{Path, PathBuf};
+use std::process::{Command, ExitStatus, Stdio};
+use std::{env, io, process};
+use tee::TeeReader;
+
+pub fn build_elf(
+    cmd: CargoCommand,
+    message_format: &str,
+    args: &Vec<String>,
+) -> (ExitStatus, Vec<Message>) {
+    let mut command = make_cargo_build_command(cmd, message_format, args);
+    let mut process = command.spawn().unwrap();
+    let command_stdout = process.stdout.take().unwrap();
+
+    let mut tee_reader;
+    let mut stdout_reader;
+    let buf_reader: &mut dyn BufRead = if message_format == "json-render-diagnostics" {
+        stdout_reader = BufReader::new(command_stdout);
+        &mut stdout_reader
+    } else {
+        tee_reader = BufReader::new(TeeReader::new(command_stdout, io::stdout()));
+        &mut tee_reader
+    };
+
+    let messages = Message::parse_stream(buf_reader)
+        .collect::<io::Result<_>>()
+        .unwrap();
+
+    (process.wait().unwrap(), messages)
+}
+
+fn make_cargo_build_command(
```

Comment:
Some of these functions used to have documentation comments. Please keep those :).

Reply:
All of the background functions have the same comments as before. I double checked the commit diff to make sure (on mobile at the moment so could still have missed one). I can add comments though for a better understanding.